sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/TableIdentifierParserSuite.scala

```diff
@@ -285,334 +290,86 @@ class TableIdentifierParserSuite extends SparkFunSuite with SQLHelper {
-    "current_user",
-    "day",
-    "distinct",
-    "else",
-    "end",
-    "escape",
-    "except",
-    "false",
-    "fetch",
-    "for",
-    "foreign",
-    "from",
-    "full",
-    "grant",
-    "group",
-    "having",
-    "hour",
-    "in",
-    "inner",
-    "intersect",
-    "into",
-    "join",
-    "is",
-    "leading",
-    "left",
-    "minute",
-    "month",
-    "natural",
-    "not",
-    "null",
-    "on",
-    "only",
-    "or",
-    "order",
-    "outer",
-    "overlaps",
-    "primary",
-    "references",
-    "right",
-    "select",
-    "semi",
-    "session_user",
-    "minus",
-    "second",
-    "some",
-    "table",
-    "then",
-    "to",
-    "trailing",
-    "union",
-    "unique",
-    "unknown",
-    "user",
-    "using",
-    "when",
-    "where",
-    "with",
-    "year")
+  private def parseAntlrGrammars(startTag: String, endTag: String)
+      (f: PartialFunction[String, Option[String]]): Set[String] = {
+    // We need to map a symbol string to actual literal strings
+    // in case that they have different strings.
+    val symbolsToNeedRemap = Map(
+      "DATABASES" -> Seq("DATABASES", "SCHEMAS"),
+      "RLIKE" -> Seq("RLIKE", "REGEXP"),
+      "SETMINUS" -> Seq("MINUS"),
+      "TEMPORARY" -> Seq("TEMPORARY", "TEMP")
+    )
+    val keywords = new mutable.ArrayBuffer[String]
+    val default = (_: String) => None
+    var startTagFound = false
+    var parseFinished = false
+    val lineIter = sqlSyntaxDefs.toIterator
+    while (!parseFinished && lineIter.hasNext) {
+      val line = lineIter.next()
+      if (line.trim.startsWith(startTag)) {
+        startTagFound = true
+      } else if (line.trim.startsWith(endTag)) {
+        parseFinished = true
+      } else if (startTagFound) {
+        f.applyOrElse(line, default).foreach { symbol =>
+          if (symbolsToNeedRemap.contains(symbol)) {
+            keywords ++= symbolsToNeedRemap(symbol)
+          } else {
+            keywords += symbol
+          }
+        }
+      }
+    }
+    assert(keywords.nonEmpty && startTagFound && parseFinished, "cannot extract keywords from " +
+      s"the `SqlBase.g4` file, so please check if the start/end tags (`$startTag` and `$endTag`) " +
+      "are placed correctly in the file.")
+    keywords.map(_.trim.toLowerCase(Locale.ROOT)).toSet
+  }
 
-  val nonReservedKeywordsInAnsiMode = allCandidateKeywords -- reservedKeywordsInAnsiMode
+  private def parseSyntax(startTag: String, endTag: String): Set[String] = {
+    val kwDef = """\s*[\|:]\s*([A-Z_]+)\s*""".r
+    parseAntlrGrammars(startTag, endTag) {
+      // Parses a pattern, e.g., `    | AFTER`
+      case kwDef(symbol) => Some(symbol)
+    }
+  }
+
+  // All the SQL keywords defined in `SqlBase.g4`
+  val allCandidateKeywords = {
+    val kwDef = """([A-Z_]+):.+;""".r
+    val keywords = parseAntlrGrammars(
+        "//--SPARK-KEYWORD-LIST-START", "//--SPARK-KEYWORD-LIST-END") {
+      // Parses a pattern, e.g., `AFTER: 'AFTER';`
```

Comment:
for `KEYWORD: "ABC" | "XYZ"`, it's one keyword or two?

Reply:
Two. As for the pattern, we need to expand "KEYWORD" to two keywords in https://github.com/apache/spark/pull/28802/files#diff-7e16f8441ff7d93ab9a9cf8e23c3df4cR310-R315
```
    val symbolsToNeedRemap = Map(
      "DATABASES" -> Seq("DATABASES", "SCHEMAS"),
      "RLIKE" -> Seq("RLIKE", "REGEXP"),
      "SETMINUS" -> Seq("MINUS"),
      "TEMPORARY" -> Seq("TEMPORARY", "TEMP"),

      "KEYWORD" -> Seq("ABC", "XYZ") <-- we need to expand it like this
    )
```

Reply:
can we build this `symbolsToNeedRemap` automatically?

e.g. we go through the keyword list, for `KEYWORD: "KEYWORD"`, just keep it in a list, for KEYWORD: "ABC" | "XYZ", we put in in `symbolsToNeedRemap`.

Reply:
hm, okay. I'll try.

Reply:
@cloud-fan Cold you check the latest commit?